src/sph-duplicate2.cc

```diff
@@ -0,0 +1,1095 @@
+        virtual ~ColDefForcedSource () {}
+
+    protected:
+
+        /// Evaluate the column
+        virtual void _evaluateColumn (std::string const & name, int colnum) {
+            if      ("deepSourceId" == name) { idxDeepSourceId = colnum; }
+            else if ("chunkId"      == name) { idxChunkId      = colnum; }
+            else if ("subChunkId"   == name) { idxSubChunkId   = colnum; }
+        }
+        
+        /// Validator for the definitions
+        virtual bool _isValid () const {
+            return idxDeepSourceId *
+                   idxChunkId *
+                   idxSubChunkId >= 0;
+        }
+
+    private:
+
+        /// Copy constructor (is disabled)
+        ColDefForcedSource (ColDefForcedSource const &);
+
+        /// Assignment operator (is disabled)
+        ColDefForcedSource& operator= (ColDefForcedSource const & lhs);
+
+    public:
+
+        int idxDeepSourceId {-1},
+            idxChunkId      {-1},
+            idxSubChunkId   {-1};
+    };
+
+    /// ForcedSource table's column definitions instance
+    ColDefForcedSource coldefForcedSource;
+
+
+    /// Write a row into a stream
+    void writeRow (std::vector<std::string> const & tokens, std::ofstream & os) {
+
+        if (opt.dryRun) return;
+
+        for (size_t idx = 0; idx < tokens.size(); ++idx) {
+            if (idx) os << "\t";
+            os << tokens[idx];
+        }
+        os << "\n";
+    }
+
+    /// The transformation map sbetween the old and new primary keys of object tables
+    typedef std::map<uint64_t,uint64_t> ObjectIdTransformMap;
+
+    /// Transformation table instances. One map is for the original (input) objects,
+    /// and the other one - for the duplicate ones.
+    ObjectIdTransformMap objIdTransformInput,
+                         objIdTransformDuplicate;
+
+    /// Objects which were found out-of-the partition box. THese objects
+    /// will not be duplicated or recorded into the output streams.
+    std::set<uint64_t> objIdOutOfBox;
+
+    /// Duplicate the next row of the chunk's Object table
+    size_t duplicateObjectRow (std::string              & line,
+                               part::SphericalBox const & box,
+                               std::ofstream            & os) {
+
+        // Split the input line into tokens and store them
+        // in a temporrary array at positions which are supposed to match
+        // the correposnding ColDef
+
+        std::stringstream is(line);
+
+        std::vector<std::string> tokens(coldefObject.columns.size());
+
+        size_t colnum = 0;
+
+        std::string token;
+        while (std::getline(is, token, '\t')) {
+            if (colnum == tokens.size())
+                throw new std::range_error("too many tokens in a row of the input Object file");
+            tokens[colnum++] = token;
+        }
+        if (colnum != tokens.size())
+            throw new std::range_error("too few tokens in a row of the input Object file");
```

Comment:
If you need to split a line into tokens then Boost has a string algorithm library which is probably much more efficient that doing it via stringstream: http://www.boost.org/doc/libs/1_60_0/doc/html/string_algo/usage.html#idm45555128601440